hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/NativeAzureFileSystem.java

```diff
@@ -915,6 +922,33 @@ public synchronized int read(byte[] b, int off, int len) throws FileNotFoundExce
       }
     }
 
+    @Override
+    public int read(long position, byte[] buffer, int offset, int length)
+        throws IOException {
+      if (in instanceof PositionedReadable) {
+        try {
+          int result = ((PositionedReadable) this.in).read(position, buffer,
+              offset, length);
+          if (null != statistics && result > 0) {
+            statistics.incrementBytesRead(result);
+          }
+          return result;
+        } catch (IOException e) {
+          Throwable innerException = NativeAzureFileSystemHelper.checkForAzureStorageException(e);
+          if (innerException instanceof StorageException) {
+            LOG.error("Encountered Storage Exception for read on Blob : {}"
+                + " Exception details: {} Error Code : {}",
+                key, e, ((StorageException) innerException).getErrorCode());
+            if (NativeAzureFileSystemHelper.isFileNotFoundException((StorageException) innerException)) {
+              throw new FileNotFoundException(String.format("%s is not found", key));
+            }
+          }
+          throw e;
+        }
+      }
+      return super.read(position, buffer, offset, length);
```

Comment:
I think if "in" is not PositionedReadable steam then we should throw exception. We shouldn't change the meaning of read() API by calling super.read().

**FSInputStream.read(long, byte[], int, int)** internally call **read(byte[], int, int)**, which is not positioned read API. This will change the current offset of a file.

Reply:
Before this patch it happen this way.
FSInputStream.read(long, byte[], int, int)
You are right that FSInputStream pread API do relative read.  But in finally it seek back to original pos
ya after this patch the 'in' will be PositionedReadable for sure.  But I think we should keep this generic way
If here 'in' is PositionedReadable make use of that pread API or else fallback to old way

Reply:
Agree